core/src/Kokkos_Array.hpp

```diff
@@ -320,6 +320,26 @@ struct Array<T, KOKKOS_INVALID_INDEX, Array<>::strided> {
       : m_elem(arg_ptr), m_size(arg_size), m_stride(arg_stride) {}
 };
 
+template <typename T,
+          size_t LN, typename LP, 
+          size_t RN, typename RP>
+KOKKOS_INLINE_FUNCTION constexpr bool operator==(const Array<T, LN, LP>& lhs, 
+                                                 const Array<T, RN, RP>& rhs) {
```

Comment:
`std::array` also doesn't allow compring `std::array`s with different numbers of elements, see https://godbolt.org/z/Y5b3fzGfq.
```suggestion
template <typename T,
          size_t N, typename LP, 
          typename RP>
KOKKOS_INLINE_FUNCTION constexpr bool operator==(const Array<T, N, LP>& lhs, 
                                                 const Array<T, N, RP>& rhs) {
```
I'm not quite sure about the `Proxy` argument, though.

Reply:
I'm not sure what's best to do.

The four versions of `Kokkos::Array` are:

```
template <class T = void, size_t N = KOKKOS_INVALID_INDEX, class Proxy = void> 
struct Array;

template <class T, class Proxy> 
struct Array<T, 0, Proxy>;

template <class T> 
struct Array<T, KOKKOS_INVALID_INDEX, Array<>::contiguous>;

template <class T>
struct Array<T, KOKKOS_INVALID_INDEX, Array<>::strided>;
```

And so the `Proxy` influences `N`. I.e., for the last two specialisations (wraps of dynamically allocated arrays), the size is set to the invalid value. If we limit `operator==` to only comparisons of equal size, I think we won't be able to compare `Kokkos::Array` that wrap a statically allocated array with `Kokkos::Array` that wrap a dynamically allocated array.

The fact that the `Kokkos::Array` can wrap a dynamically allocated array seems to be a difference with the `std::Array`. So to me, it's OK that there is also a difference in what `operator==` can do. But, if you prefer, I can update the implementation and limit it to comparisons of the same size.

Reply:
Hi @masterleinad. I just wanted to ask if you should have had a chance to think about what would be best to do? Should we limit the comparison to the same size to be consistent with the std::array as you had suggested? Or should we allow comparisons between kokkos::array with different size template parameter to allow comparisons of kokkos::array of different proxy template parameter, such as static and dynamic ones? Thanks in advance.

Reply:
Ideally, I would want to eliminate the `Proxy` template argument but that's a different discussion. Unless you have a use case where you need it, I would still prefer to follow the restrictions set by `std::array` and only provide the comparison operator if `Proxy==void`. It's easier to relax it later than to restrict it.

Reply:
OK. I've just revised it as you suggest. The comparison is now only provided if the arrays have equal size and `Proxy==void`. 

(Just a thought: in a discussion about removing `Proxy`, perhaps the possibility of replacing the statically sized `Kokkos::Array` with `std::Array` could also be explored; in our codes, `std::Array` seems to work at least in HIP and Cuda kernels, but I don't know about the other backends).